nanofi/src/api/ecu.c

```diff
@@ -0,0 +1,271 @@
+              case 'n':
+                delim = '\n';
+                break;
+              case '\\':
+                delim = '\\';
+                break;
+              default:
+                break;
+            }
+        }
+    }
+
+    int len = strlen(file_path);
+    props.file_path = (char *)malloc((len + 1) * sizeof(char));
+    strncpy(props.file_path, file_path, len);
+    props.file_path[len] = '\0';
+    props.delimiter = delim;
+    return props;
+}
+
+void on_trigger_logaggregator(processor_session * ps, processor_context * ctx) {
+
+    struct properties props = get_properties(ctx);
+
+    if (!props.file_path) return;
+
+    char delim = props.delimiter;
+
+    flow_file_info ff_info = log_aggregate(props.file_path, delim, get_offset());
+    set_offset(get_offset() + ff_info.total_bytes);
+    ff_list = ff_info.ff_list;
+    free(props.file_path);
+}
+
+void write_flow_file(flow_file_record * ffr, const char * buff, int count) {
+    FILE * ffp = fopen(ffr->contentLocation, "ab");
+    if (!ffp) return;
+    if (fwrite(buff, 1, count, ffp) < count) {
+        fclose(ffp);
+        free_flowfile(ffr);
+        flow_file_exists = 0;
+        return;
+    }
+    fflush(ffp);
+    fclose(ffp);
+}
+
+void on_trigger_tailfiledelimited(processor_session * ps, processor_context * ctx) {
+    struct properties props = get_properties(ctx);
+
+    if (!props.file_path) return;
+
+    char delim = props.delimiter;
+    FILE * fp = fopen(props.file_path, "rb");
+
+    if (!fp) {
+        printf("Unable to open file. {file: %s, reason: %s}\n", props.file_path, strerror(errno));
+        return;
+    }
+
+    char buff[MAX_BYTES_READ + 1];
+    size_t bytes_read = 0;
+    fseek(fp, curr_offset, SEEK_SET);
+
+    while ((bytes_read = fread(buff, 1, MAX_BYTES_READ, fp)) > 0) {
+        buff[bytes_read] = '\0';
+        const char * begin = buff;
+        const char * end = NULL;
+
+        while ((end = strchr(begin, delim))) {
+            uint64_t len = end - begin;
+            if (len > 0) {
+                if (flow_file_exists && ffr_delim) {
+                    write_flow_file(ffr_delim, begin, len);
+                    add_flow_file_record(&ff_list, ffr_delim);
+                    flow_file_exists = 0;
+                }
+                else {
+                    ffr_delim = generate_flow_file(instance, proc);
+                    write_flow_file(ffr_delim, begin, len);
+                    add_flow_file_record(&ff_list, ffr_delim);
+                    flow_file_exists = 0;
+                }
+                curr_offset += (len+1);
+            }
+            else {
+                if (flow_file_exists && ffr_delim) {
+                    add_flow_file_record(&ff_list, ffr_delim);
+                    flow_file_exists = 0;
+                }
+                curr_offset++;
+            }
+            begin = (end + 1);
+        }
+
+        if (!end && begin && *begin != '\0' && bytes_read == MAX_BYTES_READ) {
```

Comment:
I do not think `begin` can be `NULL`, so that check is unnecessary.
I see some inconsistency in this conditional.
We only enter it if we could read a full `MAX_BYTES_READ` from the file. But if in this buffer there has been complete lines, we will still create a partial flow_file from the remainder of the buffer.
I do not think this is a bad behaviour, but if we already to that, why can't we do the same if there has been less than `MAX_BYTES_READ`?
I do not think this causes erroneus behaviour, I am just trying to understand the rationale behind it.